### ISSUE TITLE
Add default value in project

### ISSUE THREAD

**Issue by mforets** [JuliaReach/LazySets.jl#1331]
I think it would be nice to have a default output set type in `LazySets.Approximations.project`.

```julia
juila> Z = rand(Zonotope);

julia>  project(Z, [1, 2])
MethodError: no method matching project(::Zonotope{Float64}, ::Array{Int64,1})

julia>  project(Z, [1, 2], LinearMap)
LazySets.LinearMap{Float64,Zonotope{Float64},Float64,SparseArrays.SparseMatrixCSC{Float64,Int64}}(
  [1, 1]  =  1.0
  [2, 2]  =  1.0, Zonotope{Float64}([0.25482, -1.88379], [1.17086 1.00467 -0.526207 -0.958695; -0.338528 0.680383 0.687293 1.76753]))
```

Using a lazy linear map by default sounds good to me.

CC: @kostakoida

**Comment by schillic** [comment]
I am fine with that. The implementation probably requires to add a new method.

**Comment by mforets** [comment]
Actually i think this deserves further discussion. I would argue to make some (breaking) changes:

1- `project` should return the "exact" projection, i.e. a concrete set.
2- `Projection(::LazySet, ::AbstractVector)` should be a helper constructor for the linear map of a projection where you pass the variables on which you want to project
3- Other `project` functions can be removed, e.g. those that are a composition of a lazy projection (with a `LinearMap`) and with an overapproximation.

Note: some `project` methods do already behave like 2, eg. `project(P::HPolyhedron{N}, block::AbstractVector{Int})` and `project(H::HalfSpace{N}, block::AbstractVector{Int})`.

**Comment by schillic** [comment]
`project` is originally just a helper function for `decompose`. So I suggest to rename this helper method in a first step.

**Comment by mforets** [comment]
> 1- project should return the "exact" projection, i.e. a concrete set.

This is now the case in the current development version. There is still room for improvement and specialization. For instance,

```julia
julia> project(Z, [1, 2])
HPolyhedron{Float64}(HalfSpace{Float64,VN} where VN<:AbstractArray{Float64,1}[HalfSpace{Float64,Array{Float64,1}}([-0.27016058730099896, -0.962815276711571
2], 3.234200957777295), HalfSpace{Float64,Array{Float64,1}}([0.27016058730099896, 0.9628152767115712], -0.012263323795623116), HalfSpace{Float64,Array{Floa
t64,1}}([-0.21912586356278257, -0.9756966003414508], 3.3472199350282574), HalfSpace{Float64,Array{Float64,1}}([0.21912586356278257, 0.9756966003414508], 0.
06475586071966877), HalfSpace{Float64,Array{Float64,1}}([0.753442595972694, 0.657513691548646], -0.13540947567080308), HalfSpace{Float64,Array{Float64,1}}(
[-0.753442595972694, -0.657513691548646], 2.1590202413008805), HalfSpace{Float64,Array{Float64,1}}([0.80848188787096, 0.5885210590833674], 0.00045263028336
806066), HalfSpace{Float64,Array{Float64,1}}([-0.80848188787096, -0.5885210590833674], 2.0723704048884874)])

julia> @which project(Z, [1, 2])
project(P::AbstractPolyhedron{N}, block::AbstractArray{Int64,1}) where N in LazySets.Approximations at /home/mforets/.julia/dev/LazySets/src/Approximations
/decompositions.jl:636
```
This is using the costly `constraints_list` of Zonotopes, but it would be much more efficient to use the concrete linear map here.

> 2- Projection(::LazySet, ::AbstractVector) should be a helper constructor for the linear map of a projection where you pass the variables on which you want to project

Two alternatives are to create a new type called `Projection`, or use `LinearMap`. I still think that the latter is a good idea; perhaps adding a constructor like the one proposed in this point 2, to simplify the usage. I don't know of any specialization for `Projection` that doesn't apply to `LinearMap`, so i think it is not justified to create an independent type of `LinearMap`.

> 3- Other project functions can be removed, e.g. those that are a composition of a lazy projection (with a LinearMap) and with an overapproximation.

That would be a separate issue. The API of the concrete projection should be revised at some point.

**Comment by schillic** [comment]
> This is using the costly constraints_list of Zonotopes, but it would be much more efficient to use the concrete linear map here.

Yes, (arguably after #1429) we can add specialized methods. See also #1209.

> I don't know of any specialization for Projection that doesn't apply to LinearMap, so i think it is not justified to create an independent type of LinearMap.

I liked that idea. After #1113 it should be rather easy to add that type. I would think that a projection should be simpler than an arbitrary map, but admittedly I do not have a concrete example.

**Comment by mforets** [comment]
> Yes, (arguably after #1429) we can add specialized methods.

Ok. Heuristically, for zonotopic sets the projection is cheap so we can use dispatch on `project`. Otherwise the current fallback is good, using the constraint list.

True, but we don't have meaningful specializations in mind, do we? The idea of lazy projection constructor is now https://github.com/JuliaReach/LazySets.jl/pull/1967

**Comment by mforets** [comment]
To recap, `project` now defaults to the concrete projection, and there is `Projection` for its lazy counterpart (or a type can be passed to `project`, this remains from previous versions..)

```julia
julia> using Revise, LazySets
[ Info: Recompiling stale cache file /home/mforets/.julia/compiled/v1.2/LazySets/NjrGc.ji for LazySets [b4f0291d-fe17-52bc-9479-3d1a343d9043]

julia> Z = rand(Zonotope);

julia> project(Z, [1, 2])
HPolyhedron{Float64}(HalfSpace{Float64,VN} where VN<:AbstractArray{Float64,1}[HalfSpace{Float64,Array{Float64,1}}([-0.26202182292832754, -0.965061948430937], 2.219
521279913539), HalfSpace{Float64,Array{Float64,1}}([0.26202182292832754, 0.965061948430937], 0.4556412741552611), HalfSpace{Float64,Array{Float64,1}}([0.5431835303
654625, -0.8396139900809853], 1.7058241832961047), HalfSpace{Float64,Array{Float64,1}}([-0.5431835303654625, 0.8396139900809853], 1.3661423451031878)])

julia> project(Z, [1, 2], LinearMap)
LinearMap{Float64,Zonotope{Float64},Float64,SparseArrays.SparseMatrixCSC{Float64,Int64}}(
  [1, 1]  =  1.0
  [2, 2]  =  1.0, Zonotope{Float64}([-0.7747646805967876, -0.7035141630364177], P[1.9918209311661668 1.5090667962392903; -0.5407948704013241 0.9762822435337115]))

julia> Projection(Z, [1, 2])
LinearMap{Float64,Zonotope{Float64},Float64,SparseArrays.SparseMatrixCSC{Float64,Int64}}(
  [1, 1]  =  1.0
  [2, 2]  =  1.0, Zonotope{Float64}([-0.7747646805967876, -0.7035141630364177], [1.9918209311661668 1.5090667962392903; -0.5407948704013241 0.9762822435337115]))
```
